File: 34-find-first-and-last-position-of-element-in-sorted-array/34-find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
private:
    int firstOccurance(vector<int>& nums, int target){
        int index = -1;
        int low,high;
        low = 0,high = nums.size()-1;
        while(low<=high){
            int mid = low + (high-low)/2;
            if(nums[mid]==target){
                if(0!=mid && nums[mid-1]==target){
                    index = mid-1;
                    high = mid-2; 
                }else{
                    index = mid;
                    break;
                }
            }else if(nums[mid]>target){
                high = mid-1;
            }else{
                low = mid+1;
            }
        }
        return index;
    }
    int lastOccurance(vector<int>& nums, int target){
        int index = -1;
        int low,high;
        low = 0,high = nums.size()-1;
        while(low<=high){
            int mid = low + (high-low)/2;
            if(nums[mid]==target){
                if(nums.size()-1!=mid && nums[mid+1]==target){
                    index = mid+1;
                    low = mid+2; 
                }else{
                    index = mid;
                    break;
                }
            }else if(nums[mid]>target){
                high = mid-1;
            }else{
                low = mid+1;
            }
        }
        return index;
    }
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int firsto = firstOccurance(nums,target);
        int lasto = lastOccurance(nums,target);
        return {firsto,lasto};
    }
};

File: 34-find-first-and-last-position-of-element-in-sorted-array/34-find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def firstOccurance(self,nums:List[int],target:int)->List[int]:
        ans = -1;
        low = 0;
        high = len(nums)-1;
        while low<=high:
            mid = low + (high-low)//2;
            if(nums[mid]==target):
                if(0<mid and nums[mid-1]==target):
                    ans = mid-1;
                    high = mid-2;
                else:
                    ans = mid;
                    break;
            elif(nums[mid]>target) :
                high = mid-1;
            else:
                low = mid+1;
            
            
        return ans;
        
    def lastOccurance(self,nums:List[int],target:int)->List[int]:
        ans = -1;
        low = 0;
        high = len(nums)-1;
        while low<=high:
            mid = low + (high-low)//2;
            if(nums[mid]==target):
                if(len(nums)-1>mid and nums[mid+1]==target):
                    ans = mid+1;
                    high = mid+2;
                else:
                    ans = mid;
                    break;
                ans = mid;
                low = mid+1;
            elif(nums[mid]>target):
                high = mid-1;
            else:
                low = mid+1;
            
            
        return ans;
        
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        focuurance = self.firstOccurance(nums,target);
        loccurance = self.lastOccurance(nums,target);
        return [focuurance,loccurance];